packages/z3/z3.4.8.13/opam

```diff
@@ -0,0 +1,37 @@
+opam-version: "2.0"
+maintainer: "simon@aestheticintegration.com"
+authors: "MSR"
+homepage: "https://github.com/Z3prover/z3"
+bug-reports: "https://github.com/Z3prover/z3/issues"
+license: "MIT"
+dev-repo: "git+https://github.com/Z3prover/z3.git"
+build: [
+  [ "python2.7" "scripts/mk_make.py" "--ml"]
+  [ make "-C" "build" "-j" jobs ]
+]
+
+install: [
+  [ "sh" "-c" "ocamlfind install z3 build/api/ml/* build/libz3.*"]
```

Comment:
`[ "sh" "-c" "ocamlfind install z3 build/api/ml/* -dll build/libz3.*"]`

Here needs a `-dll` to guarantee either `libz3.so` or `libz3.dylib` to be installed into `<opam-sitelib>/stublibs`.

Reply:
Thanks for taking a look. I see in the man page of `ocamlfind install` that:

> If there are files ending in the suffixes .so or .dll, the package directory will be added to the DLL configuration file ld.conf, such that the dynamic loader can find the DLL. [...]
> 
> However, if there is a stublibs directory in site-lib, the DLLs are not installed in the package directory, but in this directory that is shared by all packages that are installed at the same location. [...]
> 
> The options -dll and -nodll can be used to control exactly which files are considered as DLLs and which not. By default, the mentioned suffix rule is in effect: files ending in ".so" (Unix) or ".dll" (Windows) are DLLs. The switch -dll changes this, and all following files are considered as DLLs, regardless of their suffix. The switch  -nodll expresses that the following files are not DLLs, even if they have a DLL-like suffix.

It seems it should do the right thing if we do not specify the `-dll` option. Do you still think we need it?

Reply:
I think it's the right thing to do with the `-dll`.
There is some inconsistency between z3, opamfind, and macos.
z3 itself will make `libz3.so` on ubuntu and `libz3.dylib` on macos.
z3's ocamlbinding, which uses `ocamlmklib`, will make `dllz3ml.so` on both platforms.

We need (it makes things much simpler) to install `dllz3ml.so` and `libz3.{so|dylib}` into the `stublibs` explicitly.

I don't think `ocamlfind` handles it correctly for macos. I made this [PR](https://github.com/ocaml/ocamlfind/pull/25) there but no response yet. We need to guard it here on our own.

As for macos, it's more complex and subtle. (Please forgive my laziness to not post the reference here, but TLDR;) both `.so` and `.dylib` can be used.

Reply:
Thanks, @arbipher, I just tried running a simple test after adding the -dll option to the opam file. It works fine on macOS but on Ubuntu, it still fails with:

```
File "Z3Example.ml", line 1:
Error: Error on dynamically loaded library: /home/kartik/.opam/default/lib/stublibs/dllz3ml.so: libz3.so: cannot open shared object file: No such file or directory
```

even though the file exists in the stublibs directory:
```
~/.opam/default/lib/stublibs$ ls
dllz3ml.so  dllz3ml.so.owner  dllzarith.so  dllzarith.so.owner  libz3.so  libz3.so.owner
```

It is as if the search path is different on linux than on macOS.